torch/csrc/stable/tensor.h

```diff
@@ -0,0 +1,155 @@
+
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+#include <torch/csrc/stable/library.h>
+
+namespace torch::stable {
+
+using DeviceIndex =
+    int8_t; // this is from c10/core/Device.h and can be header only?
+
+inline void delete_tensor_object(AtenTensorHandle ath) {
+  AOTI_TORCH_ERROR_CODE_CHECK(aoti_torch_delete_tensor_object(ath));
+}
+
+// The torch::stable::Tensor class is a highlevel C++ header-only wrapper around
+// the C shim Tensor APIs. There are several goals of this class over
+// AtenTensorHandle and RAIIAtenTensorHandle:
+// 1. torch::stable::Tensor is a nicer UX much closer to torch::Tensor than the
+// C
+//    APIs with AtenTensorHandle. Under the hood we still call to these C shim
+//    APIs to preserve stability.
+// 2. RAIIAtenTensorHandle boils down to a uniq_ptr that forces the user to pass
+//    around ownership. This makes it difficult to pass one input into 2
+//    different functions, e.g., doing something like c = a(t) + b(t) if a and b
+//    both require ownership of t. Here, we use a shared_ptr accompanied with
+//    some other tricks to make our contract-based memory management much more
+//    possible.
+class Tensor {
+ private:
+  std::shared_ptr<AtenTensorOpaque> ath_;
+
+ public:
+  Tensor() = delete;
+
+  // Wrap AtenTensorHandle
+  explicit Tensor(AtenTensorHandle ath) : ath_(ath, delete_tensor_object) {}
+
+  // Wrap StableIValue
+  explicit Tensor(StableIValue siv)
+      : ath_(to<AtenTensorHandle>(siv), delete_tensor_object) {}
+
+  // Copy and move constructors can be default cuz the underlying handle is a
+  // shared_ptr
+  Tensor(const Tensor& other) = default;
+  Tensor(Tensor&& other) noexcept = default;
+
+  // Copy and move assignment operators can be default cuz the underlying handle
+  // is a shared_ptr
+  Tensor& operator=(const Tensor& other) = default;
+  Tensor& operator=(Tensor&& other) noexcept = default;
+
+  // Destructor can be default: shared ptr has custom deletion logic
+  ~Tensor() = default;
+
+  AtenTensorHandle get() const {
+    return ath_.get();
+  }
+
+  StableIValue get_StableIValue() const {
+    AtenTensorHandle handle = ath_.get();
+
+    // The following is our way of incrementing the refcount of the underlying
+    // Tensor that we point to. Why do we want this supposedly weird behavior?
+    // Because! We expect users to only need a StableIValue when they are trying
+    // to pass the Tensor into a stack-based API, e,g.,
+    // aoti_torch_call_dispatcher.
+    //
+    // A stack-based API is one that expects a stack of inputs converted to
+    // StableIValues. Our contract with any stack-based API is that the stack
+    // has ownership of its Tensor arguments. Since this torch::stable::Tensor
+    // object will likely go out of scope by the end of the user extension's
+    // local function and will thus delete its reference on the at::Tensor,
+    // we create a new AtenTensorHandle for that use case.
```

Comment:
I think another possible use case of StableIValue is to pass as input parameters to native runtime. The ownership needs to be sorted out for that use case as well.

Reply:
Okay, implemented